--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -5105,6 +5105,18 @@ static int ath10k_add_interface(struct i
 		}
 	}
 
+	vdev_param = ar->wmi.vdev_param->rx_decap_type;
+	if (test_bit(ATH10K_FLAG_ETHERNET_MODE_ENABLED, &ar->dev_flags) &&
+	    vdev_param != WMI_VDEV_PARAM_UNSUPPORTED) {
+		ret = ath10k_wmi_vdev_set_param(ar, arvif->vdev_id, vdev_param,
+						ATH10K_HW_TXRX_ETHERNET);
+		if (ret) {
+			ath10k_warn(ar, "failed to set rx decap type: %d\n",
+				    ret);
+			goto err_peer_delete;
+		}
+	}
+
 	spin_lock_bh(&ar->htt.tx_lock);
 	if (!ar->tx_paused)
 		ieee80211_wake_queue(ar->hw, arvif->vdev_id);
