--- a/drivers/net/wireless/ath/ath10k/debug.h
+++ b/drivers/net/wireless/ath/ath10k/debug.h
@@ -38,6 +38,7 @@ enum ath10k_debug_mask {
 	ATH10K_DBG_WMI_PRINT	= 0x00002000,
 	ATH10K_DBG_PCI_PS	= 0x00004000,
 	ATH10K_DBG_AHB		= 0x00008000,
+	ATH10K_DBG_FWLOG	= 0x00010000,
 	ATH10K_DBG_SMART_ANT	= 0x01000000,
 	ATH10K_DBG_ANY		= 0xffffffff,
 };
--- a/drivers/net/wireless/ath/ath10k/fwlog.c
+++ b/drivers/net/wireless/ath/ath10k/fwlog.c
@@ -1368,24 +1368,24 @@ static void ath10k_fwlog_print(struct ath10k *ar, u32 mod_id, u16 vap_id,
 	}
 
 	if (vap_id < FW_DBGLOG_VAPID_NUM_MAX)
-		printk(FW_DBGLOG_PRINT_PREFIX "[%u] vap-%u %s ( ",
-		       timestamp,
-		       vap_id, fw_dbglog_get_msg(mod_id, dbg_id,
-						 module_id_max,
-						 dbg_msg_arr));
+		ath10k_dbg(ar, ATH10K_DBG_FWLOG, FW_DBGLOG_PRINT_PREFIX
+			   "[%u] vap-%u %s ( ", timestamp, vap_id,
+			   fw_dbglog_get_msg(mod_id, dbg_id,
+					     module_id_max,
+					     dbg_msg_arr));
 	else
-		printk(FW_DBGLOG_PRINT_PREFIX "[%u] %s ( ",
-		       timestamp,
-		       fw_dbglog_get_msg(mod_id, dbg_id,
-					 module_id_max, dbg_msg_arr));
+		ath10k_dbg(ar, ATH10K_DBG_FWLOG, FW_DBGLOG_PRINT_PREFIX
+			   "[%u] %s ( ", timestamp,
+			   fw_dbglog_get_msg(mod_id, dbg_id,
+					     module_id_max, dbg_msg_arr));
 
 	 for (i = 0; i < numargs; i++) {
-		printk("%#x", args[i]);
-		if ((i + 1) < numargs)
-			printk(", ");
+		 ath10k_dbg(ar, ATH10K_DBG_FWLOG, "%#x", args[i]);
+		 if ((i + 1) < numargs)
+			ath10k_dbg(ar, ATH10K_DBG_FWLOG, ", ");
 	}
 
-	printk(" )\n");
+	ath10k_dbg(ar, ATH10K_DBG_FWLOG, " )\n");
 }
 
 static void ath10k_fwlog_parse_msg(struct ath10k *ar, u8 *data, int len) {
@@ -1410,8 +1410,8 @@ static void ath10k_fwlog_parse_msg(struct ath10k *ar, u8 *data, int len) {
 	len -= sizeof(dropped);
 
 	if (dropped > 0)
-		printk(FW_DBGLOG_PRINT_PREFIX "%d log buffers are dropped \n",
-		       dropped);
+		ath10k_dbg(ar, ATH10K_DBG_FWLOG, FW_DBGLOG_PRINT_PREFIX
+			   "%d log buffers are dropped \n\n", dropped);
 
 	count = 0;
 	buffer = (u32 *)data;
@@ -1428,14 +1428,14 @@ static void ath10k_fwlog_parse_msg(struct ath10k *ar, u8 *data, int len) {
 		if (moduleid >= WLAN_MODULE_ID_MAX)
 			return;
 
-		printk("[%s] ", wiphy_name(ar->hw->wiphy));
+		ath10k_dbg(ar, ATH10K_DBG_FWLOG, "[%s] ",
+			   wiphy_name(ar->hw->wiphy));
 
 		ath10k_fwlog_print(ar, moduleid, vapid, debugid,
 			           timestamp, numargs,
 			           (((u32 *)buffer) +
 				   2 + count));
 		count += numargs + 2;
-
 	}
 }
 
