--- /dev/null
+++ b/Makefile
@@ -0,0 +1,55 @@
+NAME	:= libpoly1305-donna
+MAJOR	:= 1
+MINOR	:= 0
+VER	:= $(MAJOR).$(MINOR)
+DNAMEM	:= $(NAME).so.$(MAJOR)
+DNAMEV	:= $(NAME).so.$(VER)
+DNAME	:= $(NAME).so
+SNAME	:= $(NAME).a
+SRC	:= $(wildcard *.c)
+HEADERS	:= poly1305-donna.h
+OBJ	:= $(SRC:.c=.o)
+CFLAGS	+= -std=c99 -Wall -W -Werror -O3 -fPIC -I.
+LDFLAGS	:= -L.
+LDLIBS	:=
+
+prefix		:=/usr
+libdir		:=$(prefix)/lib
+includedir	:=$(prefix)/include
+
+.PHONY: all clean fclean install rebuild
+
+all: $(SNAME) $(DNAMEV)
+
+$(SNAME): $(OBJ)
+	$(AR) $(ARFLAGS) $@ $^
+
+$(DNAMEV): LDFLAGS += -shared -Wl,-soname,$(DNAMEM)
+$(DNAMEV): $(OBJ)
+	$(CC) $(LDFLAGS) $^ $(LDLIBS) -o $@
+	rm -f $(DNAMEM) $(DNAME)
+	ln -s $@ $(DNAMEM)
+	ln -s $@ $(DNAME)
+
+clean:
+	$(RM) $(OBJ)
+
+fclean: clean
+	$(RM) $(SNAME) $(DNAMEV) $(DNAMEM) $(DNAME)
+
+install: $(SNAME) $(DNAMEV)
+	-@if [ ! -d $(DESTDIR)$(libdir) ]; then mkdir -p $(DESTDIR)$(libdir); fi
+	-@if [ ! -d $(DESTDIR)$(includedir) ]; then mkdir -p $(DESTDIR)$(includedir); fi
+	cp $(SNAME) $(DESTDIR)$(libdir)
+	chmod 644 $(DESTDIR)$(libdir)/$(SNAME)
+	cp $(DNAMEV) $(DESTDIR)$(libdir)
+	chmod 755 $(DESTDIR)$(libdir)/$(DNAMEV)
+	rm -f $(DESTDIR)$(libdir)/$(DNAMEM)
+	rm -f $(DESTDIR)$(libdir)/$(DNAME)
+	ln -s $(DNAMEV) $(DESTDIR)$(libdir)/$(DNAMEM)
+	ln -s $(DNAMEV) $(DESTDIR)$(libdir)/$(DNAME)
+	cp $(HEADERS) $(DESTDIR)$(includedir)
+	$(foreach header,$(HEADERS),chmod 644 $(DESTDIR)$(includedir)/$(header);)
+
+rebuild: fclean all
+
